rust/worker/src/execution/operators/merge_knn_results.rs

```diff
@@ -69,69 +76,97 @@ impl Operator<MergeKnnResultsOperatorInput, MergeKnnResultsOperatorOutput>
+                    let mut hnsw_result_user_ids = Vec::new();
+                    let mut hnsw_result_vectors = None;
+                    if input.include_vectors {
+                        hnsw_result_vectors = Some(Vec::new());
+                    }
+
+                    for offset_id in &input.hnsw_result_offset_ids {
+                        let user_id = reader.get_user_id_for_offset_id(*offset_id as u32).await;
+                        match user_id {
+                            Ok(user_id) => hnsw_result_user_ids.push(user_id),
+                            Err(e) => return Err(e),
+                        }
+                        if let Some(hnsw_result_vectors) = &mut hnsw_result_vectors {
+                            let record = reader.get_data_for_offset_id(*offset_id as u32).await;
+                            match record {
+                                Ok(record) => hnsw_result_vectors.push(record.embedding.to_vec()),
+                                Err(e) => return Err(e),
+                            }
+                        }
                     }
-                }
-                merge_results(
-                    &hnsw_result_user_ids,
-                    &input.hnsw_result_distances,
-                    &input.brute_force_result_user_ids,
-                    &input.brute_force_result_distances,
-                    input.k,
-                )
-            }
-            Err(e) => match *e {
-                RecordSegmentReaderCreationError::BlockfileOpenError(e) => {
-                    return Err(e);
-                }
-                RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
-                    return Err(e);
-                }
-                RecordSegmentReaderCreationError::UninitializedSegment => {
-                    // The record segment doesn't exist - which implies no HNSW results
-                    let hnsw_result_user_ids = Vec::new();
-                    let hnsw_result_distances = Vec::new();
                     merge_results(
                         &hnsw_result_user_ids,
-                        &hnsw_result_distances,
+                        &input.hnsw_result_distances,
+                        &hnsw_result_vectors,
                         &input.brute_force_result_user_ids,
                         &input.brute_force_result_distances,
+                        &input.brute_force_result_vectors,
+                        input.include_vectors,
                         input.k,
                     )
                 }
-            },
-        };
+                Err(e) => match *e {
+                    RecordSegmentReaderCreationError::BlockfileOpenError(e) => {
+                        return Err(e);
+                    }
+                    RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
+                        return Err(e);
+                    }
+                    RecordSegmentReaderCreationError::UninitializedSegment => {
```

Comment:
OK. Just to confirm we perform lazy initialization of the record segment in the HNSW knn operator?

Reply:
no - not lazy init, we just don't have anything to merge. compactor is responsible for init